flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/sink/BaseDeltaTaskWriter.java

```diff
@@ -74,7 +80,7 @@ public void write(RowData row) throws IOException {
       case INSERT:
       case UPDATE_AFTER:
         if (upsert) {
-          writer.delete(row);
+          writer.deleteKey(deleteSchemaProjection.wrap(row));
```

Comment:
Just curious,  if the `upsert` is enabled, then for the RowData with `DELETE` RowKind, we should also write the `key` only into equality delete file, right ?  (otherwise,  the whole row will write into the equality delete file with a projected delete schema, that should be incorrect ).

Reply:
https://github.com/apache/iceberg/pull/4364/files#diff-23f690799fd2e4a57a5854fd599491842f6f5caf9e7ae46eb0ccad2f983d1d62R95

Reply:
I think we also need an unit test to cover this case.

Reply:
@openinx  - wanted to let you know that I've been out sick since yesterday, but haven't forgotten about this.

I think in the case of `DELETE`, we'd still want to have the full row reference. But using the projected delete schema might not be appropriate.

I was not able to get the `BoundedTestSource` to work in this file (though I didn't try for very long).

I'm going to take another look at this when I've gotten a little rest, but if somebody wanted to write another unit test for this situation, that would be greatly appreciated.

cc @hililiwei if you're interested and have time to add a unit test where the source emits a `DELETE` record into a sink with `upsert` enabled.

Reply:
Overall, I do think you're right that there are some concerns about the projected schema in some cases.

I have another unit test I will push shortly that shows another (likely mostly separate) bug that exists in the current master branch. This PR doesn't fix it, but like I said it exists in master as well.

It has more to do with positional deletes, but as I've been out sick as of yesterday I would leave it for others to take a look at.

Reply:
I don't think that a DELETE row when upserting is expected. Shouldn't that be an error? We only expect INSERT rows when upsering.

Reply:
I was thinking that too initially, but the Flink documentation seems to potentially indicate otherwise. And I suppose it would make sense as there needs to be a way to remove records entirely.

https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/table/concepts/dynamic_tables/#table-to-stream-conversion

From the subsection `upsert stream`:

```
Upsert stream: An upsert stream is a stream with two types of messages, upsert messages and delete messages.
A dynamic table that is converted into an upsert stream requires a (possibly composite) unique key.
A dynamic table with a unique key is transformed into a stream by encoding INSERT and UPDATE changes
as upsert messages and DELETE changes as delete messages.
The stream consuming operator needs to be aware of the unique key attribute to apply messages correctly.
```

Key part: **An upsert stream is a stream with two types of messages, upsert messages and delete messages.**

Reply:
@rdblue @kbendick
sorry to mentions for old pr.
I wonder how to distinguish whether equality field in equality delete file was append by `DELETE` or `INSERT / UPDATE_AFTER` when scan iceberg table.

Reply:
There is no way to distinguish between why the equality delete was written. Especially as the equality delete could have been written by other engines.